### ISSUE TITLE
Add guidance about bumping the minimum required `rustc` version

### ISSUE THREAD

**Issue by danburkert** [rust-lang/api-guidelines#123]
As far as I know, this repository does not offer guidelines on when it is appropriate to bump a library crate's minimum required `rustc` version.  I've found that authors of widely used and foundational crates do not consider bumping the minimum `rustc` version to be a SemVer incompatible or breaking change ([1], [2]).  I think it would be appropriate to document what norms the Rust community expects on this issue, so that deviations from the norm can be fixed without having to re-litigate the meaning of SemVer.

My personal opinion on the matter is that bumping the required `rustc` version of a library is a backwards-incompatible SemVer change, since it requires applications depending on the library to follow suit.  This change can be every bit as painful for application authors as introducing a breaking API change (or more). The minimum `rustc` version is 'viral', and a breaking change by a foundational crates forces all other downstream crates to in turn bump their minimum version.  Cargo has good tools for introducing new features which may require a newer `rustc` version, while it has relatively poor support for resolving `rustc` version dependencies if the author of the library does not follow semantic versioning.  Thus, unless Cargo's version resolution begins taking into account required `rustc` versions, it's better for library authors to put new code requiring new `rustc` versions behind a feature flag.

[1]: https://github.com/Geal/nom/issues/500
[2]: https://github.com/alexcrichton/gcc-rs/issues/240

**Comment by KodrAus** [comment]
I think it's fair and simple to assume any change to the minimum `rustc` version is a breaking change.

What's less clear is _when_ that happens. If you happen to have a minimum version of `1.8.0` for your library then bundle in a few `?`s with other internal refactorings then you've just bumped your minimum version, possibly without even realising it.

So I think a good recommendation is to pin a `rustc` version in CI, [serde tracks `1.13.0`](https://github.com/serde-rs/serde/blob/master/.travis.yml#L7), when you want to stabilise your library and any changes that break your build on that pinned version need an appropriately breaking version bump, or gymnastics with flags.

**Comment by dtolnay** [comment]
I am less of a hardliner on this, although I hear many people think I am wrong.

Serde tracks a Rust version in CI so that we don't accidentally raise the minimum, not because we never will.

Users should not have an expectation that they can selectively pin parts of their dependency graph while receiving updates to things that depend on those parts. In Serde's case, the serde and serde_derive crates are versioned in lockstep. Serde 1.0.5 and serde_derive 1.0.5 work together, and serde 1.0.10 and serde_derive 1.0.10 work together. There shouldn't be an expectation that you can pin serde to 1.0.5 and get to use serde_derive 1.0.10. Does that constitute a "breaking change" to serde_derive? Sure, technically.

Dependencies on the standard library or language should work analogously.

**Comment by sfackler** [comment]
I'm not sure I agree that this needs to be treated as a breaking change. On one hand, you have the cost of forcing people to upgrade their rustc when `cargo update`ing. On the other, you have the cost of forcing the ecosystem to propagate new major versions throughout it.

Judging by the recent Rust community survey survey responses, ~95% of respondents were targeting 2 stable releases back or newer.

I am looking forward to picking up the `bitflags` crate's new use of associated consts. The various flags are *way* more discoverable attached to the relevant type rather than floating around the containing module. I can perform that upgrade in an API-compatible way with deprecated reexports, but if the rustc dependency bump is going to count as a breaking change, I'm just not going to do it for a year or so, probably. The pain of migrating the ecosystem forward across that breaking change is just too painful. rust-openssl 0.9 was released last November and people are *still* migrating off of the older releases. There are 1-2 issues opened per week against the rust-openssl repo that boil down to "upgrade to 0.9".

**Comment by WiSaGaN** [comment]
Usually I would expect bumping the minimum `rustc` version should not happen in bugfix release, e.g. 1.0.0 to 1.0.1, but certainly permissible when we have 1.0.0 to 1.1.0 as long as it is advertised as such.

**Comment by BurntSushi** [comment]
I have some thoughts on this, but I think that at a minimum, we should be strongly recommending that folks include a specific version of Rust in their CI builds. This doesn't imply any particular semver policy, but does at least make bumping the minimum Rust version a conscious change. As someone who has tried to play the conservative side of this fence (semver bump for minimum rustc bump), I will say that having a pinned Rust version in a CI config makes the **research task** of determining the release policy of your dependencies much easier, and typically instantly clarifies the choices I actually have available to me.

I've been trying to straddle this line for a long time now myself, and it is very hard to do so. The hardest part about this problem is that there are *significant* reasons for choosing either side of this, and I think those reasons have already been highlighted in this thread. Another difficult aspect of this problem is that since there is no consensus on whether a minimum rustc version bump is actually a semver breaking change or not, you end up with an ecosystem where one needs to be aware of the release policy for each dependency of your code, because that policy is not necessarily communicated through version numbers. Even worse, this doesn't just apply to direct dependencies, but to all transitive dependencies too.

I'd like to collect a few of my experiences with this topic.

* There is actually an RFC on this topic which I think basically says that rust-lang crates should support the previous two stable releases, but otherwise, there's no concrete policy. The RFC contains a discussion with a variety of opinions: https://github.com/rust-lang/rfcs/pull/1619
* I asked about the release policy of clap here: https://github.com/kbknapp/clap-rs/issues/740 --- Opening this issue led to a policy where a minimum rustc bump was considered a minor breaking change, so that dependents could use the `~x.y` syntax in their `Cargo.toml` and be guaranteed that a patch release will never bump the minimum Rust version.
* Unfortunately, the new release policy of clap led to @joshtriplett opening this bug against ripgrep: https://github.com/BurntSushi/ripgrep/issues/271 --- Basically, the `~x.y` syntax prevents ripgrep from compiling with newer versions of clap, *even if it could*. The interesting point brought up in this thread by @joshtriplett is the questioning of whether ripgrep itself should require a minimum Rust version. The thinking here is that the ability to compiler ripgrep `master` on a distro's (potentially) ancient version of Rust isn't particular valuable in and of itself, *but only if that distro enables packaging Rust applications easily*. In that sense, folks should be able to just install ripgrep from their distro and not need to compile it.
* @alexcrichton also notes issues with `~` dependencies: https://github.com/kbknapp/clap-rs#warning-about--dependencies
* I did eventually change the `clap` dependency to remove the `~` dependency: https://github.com/BurntSushi/ripgrep/blob/256aeb55468d57a179447c8b2cd64ada998987c5/Cargo.toml#L35 --- As a result, I almost never run `cargo update`, but rather, `cargo update -p {crate}`.
* Maintaining a minimum Rust version in the face of dependencies without that same policy is a maintenance burden: https://github.com/kbknapp/clap-rs/pull/967 --- and that in turn sometimes has unintended consequences: https://github.com/kbknapp/clap-rs/issues/1003
* A more circumspect interaction with the `encoding_rs` crate, where the outcome was this: https://github.com/hsivonen/encoding_rs#rust-version-compatibility --- Even getting [specific pinned versions in the CI config](https://github.com/hsivonen/encoding_rs/blob/aed29b8c92bcc7e29a3e34b9358d501f34ec2e33/.travis.yml#L3-L4) was a huge win, but the policy here is, in short, "probably like clap's, but no promises."
* Regardless of what ideals we all want to believe, users will try and use their distro's version of Rust to compile Rust applications, and when it fails, they will get frustrated: https://github.com/BurntSushi/ripgrep/issues/302

I think that's it as far as my own experiences go. It has been a struggle, but sometimes I wonder whether it is a natural struggle or not. This is why I think making the release policy of a crate discoverable is a first good concrete step we can take that will improve the situation for everyone, and I think it should be uncontroversial. (I think a minimal release policy is putting a pinned Rust version in a CI config.)

My own personal views on this have historically been a little hardline in the sense that I thought it was a semver breaking change. But I've grown increasingly soft on that particular issue. The end result is that I am intensely ambivalent. Therefore, I wind up on the conservative side of things because I'm not sure what to do. That is, all of my crates are pinned to a specific Rust version and I only ever bump the minimum Rust version with a semver version bump.

My conservative stance on this is what has been preventing me from pulling the trigger to release regex 1.0, for example. I *really* want to use SIMD optimizations in regex, but if I release regex 1.0 now, then under my conservative policy, I wouldn't be able to use SIMD unless I either released regex 2.0 or put the support behind a feature flag that is disabled by default. (At least, I think I could do that.) Either choice is a huge pain and not ideal at all. I suppose we shouldn't necessarily be afraid of making many semver version bumps ("so what if I release regex 2.0"), but every time I do that, some portion of the crate ecosystem is going to build both regex 1 and regex 2 into their crates, which aside from being just plain weird is going to hurt compile times quite a bit!

I have been considering what I think of as the clap compromise for regex though, because I also think it's weird to be holding back from regex 1.0 because of this. In particular:

1. regex 1.x promises API compatibility.
2. regex 1.x.y promises Rust compiler compatibility.

Invariably, this results in the same suggestion that clap makes: if you care about Rust compiler compatibility, then use `~1.x` in your `Cargo.toml`, which in turn creates other problems I've highlighted elsewhere. I think my own personal take is that this is a decent compromise, and that I'd prefer the problems associated with the clap compromise over the problems associated with "be conservative" *and* the problems associated with "move fast and break things."

**Comment by cuviper** [comment]
I still wish that a crate's minimum rustc could be expressed to cargo, as well as minimum std and cargo if we don't conflate all of these.  Then cargo could just use that for dep-solving, pruning out versions from the registry that are too new for the toolchain in use.  That still requires crates to be diligent about declaring their requirement, hopefully with CI to prove it, but it would be better than the status quo.

There have been a few RFCs started along these lines, and I admit I just sat back on the assumption that things were headed on the right track.  Can anyone summarize what's happened with that?

**Comment by joshtriplett** [comment]
@cuviper Rough summary: any kind of version specification depends on new features in Cargo and `Cargo.toml`, so these efforts blocked on things like Cargo unstable features and versioning of the Cargo schema itself. I think https://github.com/rust-lang/rfcs/pull/2182 is the next step there, and once we have that we can introduce a means of handling rustc.

**Comment by BurntSushi** [comment]
> For things like SIMD optimizations that don't affect the public API, there is a third option. You can check the rustc version in a build script and set appropriate CFGs to enable or disable the SIMD codepaths automatically. This does mean that you now have two implementations of a bunch of stuff, but that may already be the case if you're specifically targeting AVX or whatever.

Yeah that's what I meant by adding a feature. It's definitely a pain but yeah that might be what I end up doing.

**Comment by Ixrec** [comment]
It does seem like a proper solution to this isn't possible without help from cargo, but it also seems like we can all agree that bumping the minimum rustc version should at least be a minor version bump, and I doubt that's going to change even when cargo has a great, stable solution for this. So for the purposes of the API guidelines, it seems safe to recommend that everyone:
- explicitly test a specific rustc version in their CI
- make that version easy to find in your documentation
- never bump the minimum rustc version in a patch release

Anything beyond that is probably overkill for a typical crate, but this is enough to make upgrading to a more elaborate policy easy if your crate ever becomes atypically popular, and it's enough to avoid putting an undue research burden on anyone that does need to support specific rustc versions.

-----

Actually, I just noticed the current guidelines don't talk about CI or versioning at all. Are those considered out of scope?

**Comment by kbknapp** [comment]
I would personally like it if more crates were willing to release >1.0. In my opinion many crates are afraid to release 1.0 for fear of never being able to go beyond. Much of what is being discussed here could be alleviated if crates were more willing to bump the major.

I believe a major hesitation to many popular crates releasing a 1.0 is the fear of "getting something wrong" or not being able to take advantage of rustc advances by not being able to go beyond 1.0 as similarly described by @BurntSushi 

I'm hoping as the "core" Rust crates start going beyond 1.0, it will become more acceptable, but don't see it happening without the "core" or very popular libraries doing it first as they generally set the tone and guidelines. Personally, I also think the, "Never a Rust 2.0" is partially contributing to this phenomenon.

This is by no means a call for crates to quickly release a 1.0 followed immediately by a 2.0. But I do wish more crates were willing to bump the major as issues like what's being discussed evolve naturally. I believe if we had more willingness to bump the major, this would probably a be a non-issue.

I'm also just as guilty, as even though clap went 2.0 quite quickly, I've been extremely hesitant to go 3.0 for fear of getting "too far ahead of the norm, and therefore something is wrong" which is admittedly a little silly. As mentioned above, I don't think the `~x.y` should be the defacto solution, and perhaps create more problems than they solve. As @BurntSushi said, there are *very valid* reasons for both sides of the argument. 

For now though, I see the above three bullet points mentioned by @lxrec as being a better policy than none at all.

**Comment by est31** [comment]
A bump of the required rust version should be treated as a a breaking change of your library in the sense of semver. It breaks the `cargo update` workflow for anyone who is not always on latest Rust stable, and this is not good.

Is it enough if you bump the minor version? No! That means that anybody who has caret requirements *anywhere* inside their dependency chain might get their `cargo update` workflow broken, and caret requirements are definitely too popular inside the rust community.

If one day we might be getting a "minimum rust version required" field in Cargo.toml, and `cargo update` gets changed to not update a dependency if the newer version has a too new required rust version, I think that requirement can be loosened.

**Comment by dtolnay** [comment]
@est31 I strongly disagree with any statements that a bump of required Rust version must always be released as a breaking change. Whatever advice we end up giving here will be primarily a discussion around evaluating tradeoffs, not a hard line.

If I have a library that miraculously compiles on Rust 1.3.0 and is used by thousands of downstream crates and I discover a performance problem that would be fixed by using [`Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw), my options are:

1. Live with the poor performance for the rest of forever,
2. Bump the minimum Rust version to 1.4.0 in a breaking change, forcing [months of ongoing breakage](https://github.com/servo/servo/issues/8608) as a thousand libraries upgrade,
3. Write unsafe code to do it myself (a future version of Rust would be free to break my unsafe implementation and cause memory unsafety),
4. Freely use `Box::from_raw` in a minor version. Zero people in the world still care about 1.3.0.

It is pretty clear to me which of these I would do and lose no sleep over.

**Comment by est31** [comment]
@dtolnay This is an argument of the form "if I remove this public function nothing bad will happen because who really uses it and doing a breaking change would require people to update".

I really hate how staying up to date on the current crates.io ecosystem is *unusable* if you have installed your rustc copy from your distro or are pinning your rust compiler for some other reason. I don't say that people should not rely on Rust `1.4.0` but at least don't break existing projects.

If you are on an older compiler you aren't even getting compilation errors like "sorry, but your Rust compiler is too old", the errors are more often of the form "function not found" or "this feature is unstable".

**Comment by est31** [comment]
And libc is literally inside the top two of most directly depended upon crates, so it is an exceptional example.

**Comment by dtolnay** [comment]
It is an argument of the form "evaluate the tradeoffs and take the least bad approach rather than adhering to dogma."

**Comment by kbknapp** [comment]
An issue with bumping the major upon increasing the minimum required Rust that I haven't seen mentioned yet is the affects of signaling "minor breaking changes"

With Rust having a very quick 6 week release schedule, it's not inconceivable that a library maintainer may want to take advantage of new features every few Rust releases (or at max every release).  Speaking from experience, there *always* seems to be that new hot feature I can't want to be stabilized. If it's advised to bump the major, it's conceivable that libraries could have very rapid major bumps. 

Library *consumers* have no way to know if a bump of a major is a *real* breaking change (non additive change of API surface, major shift of how to use, etc.) from things like minimum required Rust. I believe this would lead to more uses of `version = "*"`. This system incentives staying on old `rustc`s, or never releasing 1.0.

There is no way to say, "I'm using the latest stable and don't care about minimum required Rust changes, but I *do* care about breaking changes." You're effectively forced into the current `~x.y` scenario, without the ability to opt out.

I'd be curious to know if that scenario is any different to a package maintainer (Linux, macOS Homebrew, etc.). @joshtriplett do you have any insight for such a scenario and if it'd be any different from the `~x.y` issues (e.g. library releases many major bumps and applications/consumers are spread across a large range of majors)?

**Comment by joshtriplett** [comment]
@kbknapp For packages of Rust crates within a distribution, it's helpful to get an indication when the minimum Rust version has changed, and once we have metadata for that then we can track that in package dependencies. However, any given library crate should only get uploaded once the distribution has a sufficiently new rustc and cargo, and once it does the dependency won't matter, so this is mostly a matter of helping to make sure that gets checked before uploading.

**Comment by est31** [comment]
> Library consumers have no way to know if a bump of a major is a real breaking change (non additive change of API surface, major shift of how to use, etc.) from things like minimum required Rust.

When updating to a new major version I already now first just update it and try whether the build still works. If it does it is a good indication that nothing major changed in the API.

@kbknapp it seems that your point (as well as my previous point about `cargo update` breaking if you just allow minor version upgrades) will get moot once we can actually track required version information via the metadata. Because then one doesn't have to require that the required version gets bumped only with a major semver release (as cargo update can take the metadata information into account). So see my proposal only scoped to libraries that do not track the required version.

Btw, as for verifying the required version I think the best idea is to check whether the currently running compiler matches the required version at package time. We already do a compilation here to verify whether the crate compiles. A more sophisticated variant would be to optionally use rustup to possibly download the specified minimum version and then do the build with that version. It can be made really convenient and the same time really accurate (unless you upload to crates.io via CI, or you implement some other mechanism that automatically transports the build status for the minimum required version from CI to the computer where you do the packaging, CI is not really accurate in ensuring that the version specified in the metadata actually can still build the project).

**Comment by mgeisler** [comment]
@dtolnay, you wrote:

> If I have a library that miraculously compiles on Rust 1.3.0 and is used by thousands of downstream crates and I discover a performance problem that would be fixed by using Box::from_raw, my options are:

and listed options 1 to 4. Would there not be an option 5:

5. Write conditionally compiled code that takes advantage of the new `Box::from_raw` function. Enable this code conditionally at compile time based on the version of `rustc` being used.

With that option, you would be respecting the semantic versioning (and thus keep the expectations of your downstream dependencies) _and_ you would get to make things better/faster/easier for users with newer compilers. This is the same as what @sfackler suggesed doing about the SIMD support.

So to that end, maybe the focus should be on enabling such cases? Make it super easy and low-effort to opt-in to new features in the code. I don't think it's possible today to do conditional compilation with `#[cfg]` based on the `rustc` version?

That is, I'm imagining writing something like this:
```rust
#[cfg(semver_compat(rustc_version, "< 1.4.0"))]
fn make_me_a_box() -> Box<Stuff> {
    // slow way...
}

#[cfg(semver_compat(rustc_version, ">= 1.4.0"))]
fn make_me_a_box() -> Box<Stuff> {
    Box::from_raw_parts(...) // yay, the fast way
}
```

That is, make it easy to swap out the implementation of a function
* without having to declare new Cargo features
* without having to use a build script

This way the overhead of supporting new APIs requires less ceremony — you pay with some internal conditional logic. Importantly, downstream users won't have to see or deal with with this complexity, they only get the benefit. Their code won't suddenly stop compiling — and if they use a newer compiler, they'll see the code become faster.

**Comment by BurntSushi** [comment]
@mgeisler That would be nice. I believe you can achieve something similar today with `build.rs` shenanigans. IIRC, there was pushback on adding version specific cfgs (they have been brought up before), but I can't remember the specific reasoning.

**Comment by nagisa** [comment]
Back-referencing [my thread](https://users.rust-lang.org/t/compiler-version-support-breakage-in-semver/14590) describing real-life breakage that was reported to me.

All I’d like to point out is that CI won't help with detecting breakage for crates which aren’t routinely running CI (i.e. maintainership mode, rare commits, etc.).

Other than that it seems to me that some ability to specify rustc versions supported by the crate in Cargo.toml will be necessary. The `rustc_version`-like cfg could be implemented as a procedural macro thing and doesn't really need explicit compiler support.

**Comment by mgeisler** [comment]
> @mgeisler That would be nice. I believe you can achieve something similar today with build.rs shenanigans. IIRC, there was pushback on adding version specific cfgs (they have been brought up before), but I can't remember the specific reasoning.

@BurntSushi Yeah, I cannot imagine that it hasn't been suggested before :-) — this kind of inline feature detection is basically what people have been doing for decades with Autoconf for languages like C. The Python ecosystem also makes liberal use of such "seamless degradation" by importing missing libraries for older versions of Python. In my opinion, it works quite well.

I haven't published many Rust libraries, but I've already noticed that my code breaks from time to time, even when I don't make changes. It breaks because of my dependencies which bump the required `rustc` version. It's not very satisfying to see Travis fail from one run to another with no changes to my code and no changes to the compiler version. A community that prides itself on using semantic versioning should not accept this.

With a Rust release every 6 weeks, having support for the most recent stable versions means having support for ~3 months. That seems extremely short to me. Compare this with a language like Python where good Python libraries also support at least the 2-3 most recent recent releases — but Python releases are roughly one per year.

**Comment by kbknapp** [comment]
How does everyone think epoch's will play into this guidance? I haven't followed the epoch discussion too closely, so it's a genuine question.

**Comment by est31** [comment]
@kbknapp from the epoch RFC, they are designed to be private choices of the crate. AKA crates of different epochs can use each other. A change in the epoch should have no effect on users in most cases, unless you maybe need to avoid using keywords as identifiers or something... idk. So the way epochs affect this question seems to be solely from the standpoint of any other feature of rustc: if you use that feature and the current minimum rustc version doesn't support it, you have to bump the minimum rustc version.

So epochs don't change anything about the need to track the minimum rustc version in Cargo.toml.

**Comment by justinturpin** [comment]
I have two cents and I'd like to put them in.

I believe that requiring a new minimum rust version should not be considered a breaking change. My reasoning is twofold:

- I believe that Semver, which Rust and its libraries have decided to go with, is only concerned with public API changes, and that the compiler version does not constitute affecting the public API. It's possible that you could interpret Semver to include compiler versions as the public api, but I believe the _spirit_ is that it's only concerned with the API, ie. the code.
- Consider if it was a breaking change - major version bumps now mean two very different things to the developer. In one case, an actual API change occurred and they must change their code accordingly, and in the other they simply need to update their Rustc version and should not change any code.

This is a bit of a pain point for Rust because its a young language that's being rapidly iterated on, but I think the dust will settle and it doesn't change what a breaking change is meant to signify. I agree the rustc compiler version requirement should be described somewhere, just not here. Python's setup.py has a spot for minimum required version, I'm sure lots of other package managers do too.

**Comment by matklad** [comment]
Hi!

I've realised that I feel strongly about this issue and that I, unlike most of the commentors here, think that bumping a required `rustc` version should **never** be a major semver bump. Note that I don't maintain any high profile rust libraries, so my argument comes from fist principles and though experiments, and may not stand well against the reality :-)

The crucial observation is bumping the minimal supported `rustc` version (msrv from now on) **is not** viral. 

Consider package `foo 1.0.0`, which depends on `bar = "1.0.0"`, such that msrv for them is `1.30`. Suppose now that new `bar 1.1.0` is released, which bumps msrv to `1.32`. 

I argue that msrv for `foo` is still `1.30`. Indeed, any application, that uses `foo`, can be compiled with `1.30`, if it has `bar` locked to `1.0.0` in the lockfile. To change `foo`'s msrv, one need to change the line in `Cargo.toml` from `bar = "1.0.0"` to `bar = "1.1.0"`, as that would indeed prevent any application, which depends on `foo`, to use `rustc 1.30`. More formaly, msrv for the library is max of msrv for library's own code and msrv of **minimal** versions of its dependencies.

Libraries don't actually know which `rustc` and which specific minor versions of dependencies would be used. It is for the application (leaf crate) to decide. By not bumping major version after bumping msrv, libraries actually give the most flexibility to the application. In the `foobar` example, the application which uses `foo` and, transitively, `bar` may lock `bar` to `1.0.0`, if it needs to be compatible with `rustc 1.30`, or it may lock bar to `1.1.0`, if it is fine with requiring `rustc 1.32` and wants to take advantage of `bar`'s recent improvements (perhaps `1.1.0` added super-fast SIMD optimizations :-) ). Note that `foo` does not care at all which minor version of `bar` the application ends up picking.

In contrast, major version bump significantly decreases flexibility. Suppose that, instead of `bar 1.1.0`, `bar 2.0.0` was released, which only bumps msrv, and otherwise is API compatible with `bar 1.0.0`. This puts `foo` in a peculiar position, because now it must choose which version of `bar` to pick, `1.0.0` or `2.0.0`. Neither choice is better than other, because `1.0.0` has lower msrv, but `2.0.0` is, say, faster (because SIMD). And the application either becomes locked into `bar 1.0.0`, or `bar 2.0.0`, or, quite probably and worse, pulls *both* versions of `bar` from different dependencies. 

So, that's it for the theory. In practice, library authors need to run CI to verify msrv, and, if done naively, it'll break with suggested strategy, because Cargo always picks the latest minor versions. The simplest way to fix that is, for msrv part of CI, to use lockfile even for libraries. An even better solution should be available soon: nightly cargo has a `--minimal-versions` unstable flags, which forces the resolution to prefer lowest minor versions. The original purpose of this flag was to check, during CI, that minimal versions of dependences, specified in Cargo.toml, are in fact correct (that is, to verify that you don't depend on features, introduced in `spam 1.3.0`, if you have `spam = "1.2.0"` in toml). I think this flag would also work perfectly to check that msrv of you library is what you think it is. 

For applications, we really need to poke some Cargo folks with sharp sticks, so that they teach Cargo to generate lockfiles using only packages with specific msrv. That is probably not going to happen soon, so for the time being `cargo update -p foo` seems like the best strategy. Note, however, that it is perfectly fine *for applications* to use `~` and `=` requirements in `Cargo.toml`, so `cargo update` + incremental tilde or eq pinning of offending deps in *Cargo.toml* might also be a viable strategy. 

With all that in mind, I think the following are some reasonable guidelines:


For libraries:

* Never bump major version to only bump msrv.

* Never restrict the upper version of your dependencies to maintain msrv, because upper versions don't affect msrv by matklad's definition of msrv.

* Have CI setup running `cargo test`, using stable compiler, *without* lockfile. This job should be run daily, because it may fail if some dependency publishes bad version.

* Have a documented msrv and a CI setup running `cargo test` with it. This setup should use either a lockfile, or the `--minimal-versions` flag. 

* For any CI setup, which uses a specific compiler version, use pinned dependencies (lockfile or `--minimal-versions`). 


For binaries:

* Publish lockfile to maintain msrv.

* Use CI to verify msrv.

* Consider using `=` or `~` in Cargo.toml, to make generation of new lockfile easier. 

* Be careful when updating dependencies, use `cargo update -p` and check msrv afterwards.

**Comment by est31** [comment]
> For applications, we really need to poke some Cargo folks with sharp sticks, so that they teach Cargo to generate lockfiles using only packages with specific msrv.

In order for Cargo to be able to do that, you'd need to specify the msrv in Cargo.toml... with rustup integration cargo could even download the msrv on cargo publish and check the crate before publishing. I think the guidelines saying that you should do this would be a good compromise.

**Comment by gnzlbg** [comment]
> The simplest way to fix that is, for msrv part of CI, to use lockfile even for libraries. 

This sounds like the key guideline to me. Is the lockfile distributed as part of `cargo publish` ?

**Comment by matklad** [comment]
>This sounds like the key guideline to me. Is the lockfile distributed as part of cargo publish ?

The lockfile of a library has no effect on the downstream crates. The guideline was specifically about CI of the library itself, and it would be obsolete when `--minimal-versions` flag is stable. 

At present, we don't include lockfiles in published crates, however there's an unstable feature to do this, as it is useful for binaries to make sure `cargo install foo` always picks the same dependencies.

**Comment by anderejd** [comment]
I find the resistance to upgrading ones stable rustc version very interesting since it's very different from my own perspective. Would it help guide this situation if a list of reasons to resist upgrading rustc was compiled?

I have a hard time figuring out why anyone would avoid upgrading when it's just a `rustup upgrade` away.

**Comment by gnzlbg** [comment]
Well, the main reason to resist upgrading Rust is that upgrading the std library is a breaking change, and since there is no way to specify a std library version in the Cargo.toml, there is no way to defend against this.

For example, if we add a trait method from `Itertools` to `Iterator`, basically every crate using the `Itertool` trait method will break in the next Rust upgrade. 

And that's the "best" case. The worst case is when we add something to the `std` library that silently changes the behavior of Rust programs, like an inherent method to any type in the `std` library API.

**Comment by BurntSushi** [comment]
Note that the reasons for keeping tabs on Rust versions aren't just technical. They are also social:

* Some environments strictly control what version of software that is used. This doesn't in and of itself imply a certain semver strategy, but I think it does at least help to have a coherent story about what version of Rust your code requires. Even being able to know that requires some work.
* Some folks use distros with older versions of the Rust compiler. If you want to cater to those folks and let them compile your code, then you need to manage the version of Rust that your code requires.
* In general, as a maintainer, I want to have a clear and unambiguous statement about which version of the Rust compiler my code requires to compile.

The various semver heuristics are one way of managing this, but there are other ways (like making Cargo.toml aware of the Rust version). Others have pointed out potentially serious flaws with the semver approach. So there's no obvious path forward here.

**Comment by anderejd** [comment]
I don't know much about working in environments with strict rules on what software are allowed so I can't really contribute to that part of the discussion.

For crates.io I think it would be doable to run a big batch job of `cargo build` tests using all rustc versions and tagging each crate + crate version with compatible rustc versions. Maybe it would and should require some reasonable limit on rustc versions to test with, starting with the latest and going back a couple of stable releases.

**Comment by newpavlov** [comment]
In my opinion right now bumping the minimum required `rustc` version should be considered a breaking change (arguments already listed in the discussion above), but it should be only a temporary solution and we really should push for `rustc` field in the `Cargo.toml`, which will be used by cargo to select an appropriate crate versions. After we will get such capability, I think it will make sense to recommend for post-1.0 crates to bump minor version after changing the minimum `rustc` version. This will allow to easier publish fixes for serious bugs for crate versions which target older compilers.

**Comment by dtolnay** [comment]
Here is a crazy/wonderful approach <sub>that requires https://github.com/rust-lang/cargo/issues/5499.</sub>

Suppose Serde 1.0.x wants to bump the minimum required compiler version from rustc 1.13 to 1.29.

1. We publish Serde 2.0.0 with an implementation that requires rustc 1.29 (bear with me).
2. We go back to the code of Serde 1.0.x and add a [**semver-trick**](https://github.com/dtolnay/semver-trick)**\*** dependency on `serde = { version = "2.0", optional = true }`. This is unlike the typical semver trick dependency in that it is an optional dependency! <b>*</b><i>you need to read this link or else nothing I write will make sense.</i>
3. Still in Serde 1.0.x, we add a build.rs script which uses [`version_check`](https://crates.io/crates/version_check) to detect when we are being built by rustc >= 1.29. If so, print the magic line that turns on the optional dependency on 2.0.
4. In lib.rs of Serde 1.0.x we detect `#[cfg(feature = "serde")]` and if so, do the usual semver trick re-export of all the contents of 2.0.
5. If `#[cfg(not(feature = "serde"))]` then we provide everything as usual, everything as it exists in the 1.0.x series with a compiler requirement of 1.13.
6. Publish as the next release of 1.0.x series.

There are many steps but this is a modest amount of work for a significant library. The end state is:

- Anyone on old compilers observes no change. Dependencies on `serde = "1.0"` continue to provide all the same APIs. You cannot have 2.0 in your dependency graph because your compiler does not meet its minimum version requirement.
- On new compilers you get a semver-trick relationship where `serde = "1.0"` and `serde = "2.0"` are synonymous.

People who argue *rustc requirement bumps are a breaking change* are happy because their `serde = "1.0"` dependency can never pull in code that does not compile on their compiler version.

People who argue *rustc requirement bumps are not a breaking change* are happy because they get to use new compiler features without forcing an upgrade treadmill of major versions on their downstream users.

**Comment by retep998** [comment]
>Still in Serde 1.0.x, we add a build.rs script which uses version_check to detect when we are being built by rustc >= 1.29. If so, print the magic line that turns on the optional dependency on 2.0.

This shouldn't work as far as I know because whether a given dependency is enabled or disabled is resolved *before* build scripts are run.

**Comment by dtolnay** [comment]
Slight modification that no longer requires https://github.com/rust-lang/cargo/issues/5499: we could have 1.x unconditionally depend on 2.0 and add a build script to 2.0 as well, which sets a feature on old compilers that turns the entire 2.0 crate into an empty crate.

**Comment by luser** [comment]
I don't have strong feelings about the semver question here, but as someone maintaining [a decent-sized Rust application](https://github.com/mozilla/sccache/) I frequently run into situations where I update a dependency or add a new dependency and need to bump the minimum Rust version I support (which I test against in CI), but I find it virtually impossible to figure out what the actual required version *is*! It's easy to see that a PR breaks compiling against the pinned Rust version in CI, and often the error message will even name a specific feature that's unstable in that version, but there's no good way (AFAIK) to figure out what version stabilized that feature. (In other cases the error is not as clear and it's even harder.)

**Comment by BurntSushi** [comment]
@luser Yeah, I run into the same problem. This is why I've been advocating that everyone put a specific Rust version in their CI config so that others can go and look at it and know what version the project intends to be able to compile on. I think that is an uncontroversial and unambiguously good step forward, and in particular, does not require everyone to agree on the semver ramifications---folks are free to bump the minimum Rust version as much as they want, they just have to do so explicitly.

I think more and more projects are doing this, but unfortunately, many still aren't. At that point, you're right, it's a research task to figure out the minimum Rust version. There's probably a script someone could write that does a binary search for you, but this requires having all intermediate versions of Rust installed. (rustup makes this easy, though, may take up a bit of disk space!)

**Comment by dwijnand** [comment]
(See also https://github.com/japaric/trust/issues/103 about running CI with a rust-toolchain file present)

**Comment by newpavlov** [comment]
See rust-lang/rfcs#2495 for a proposal which will allow authors to specify minimum supported Rust version (MSRV), which in combination with additional constraints in dependency versions resolution algorithm resolution algorithm will make changing MSRV backwards compatible change.

**Comment by gnzlbg** [comment]
@luser same here, although I must say that while this is pretty annoying when it happens, it doesn't happen often enough to me to be a continuous annoyance. As in, when it happens, I bisect the version required, and don't have to worry about it again for a couple of months.

**Comment by matklad** [comment]
An interesting idea [from @Eh2406](https://github.com/rust-lang/cargo/issues/5657#issuecomment-404950336) for a CI setup that should work today. 

To test MSRV, one can use

```
cargo +nightly -Z minimal-versions generate-lockfile && cargo +MSRV test
```

Caveats:

* your crate probably won't build with `minimal-versions` right off the bat, see [1](https://github.com/rust-lang/cargo/issues/5657) and [2](https://blog.illicitonion.com/2018/06/rust-minimum-versions-semver-is-lie.html) for the list of problems you may face.

* `-Z minimal-versions` itself is unstable.

**Comment by alexheretic** [comment]
I'm using [latest stable as MSRV](https://gist.github.com/alexheretic/d1e98d8433b602e57f5d0a9637927e0c) for the crates I maintain. Hopefully better tools will appear to handle this for everyone more gracefully.

**Comment by Stargateur** [comment]
TL;DR: Strong opinion; have better error message; not a breaking change

---

Like others said, the main problem is that you don't get a nice error message about it. This would remove all the pain for my personal use of Rust. 

I don't think this should be considered as a change:

rustc is not gcc, we all? use rustup that allow to install and update rust on every common OS. People don't need to wait 1 year a .dep from debian stable to update rustc. I agree some project could have some trouble updating rust, but this should be settled by freezing the specific version of each dependency. Let's imagine ~5% of projects and I think it's much less, would have trouble to keep rustc up-to-date. That mean ~95% of users will "pay" for these ~5%, I don't think it's worth it when there is already an easy solution for these ~5%. (Also, I doubt these ~5% want to have a lot of dependency)

Semver is about API, it's doesn't have a way to express compiler version. rustc version is a meta dependency, this is not possible to include it in semver. "edition" in cargo should be enough, but in my opinion, it's clearly a way to hide a Rust 2.0. I mean, maybe I'm totally wrong, but between 2015 and 2018 edition, there is obvious breaking change. Ok, have edition make this smooth but if you try, for example, the diesel get started guide, you will have a cargo.toml with edition = 2018 but the guide didn't expect that, some code in the guide will not compile because diesel example use `self` in use statement. Also, you are forced to use "macro_use" even if you want to use 2018 edition because you can't import `table` macro.

2018 edition despite all these changes, I never saw someone complaining about update their code. I always saw, "wow nice this is so much better", I think Rust is too afraid of being seen as an unstable language because of pre-1.0. But Rust dev are in the majority not afraid of little change, on the contrary the community always asks to improve Rust. I agree Rust must not change like pre-1.0 but there is a big difference between change everything every month and change little thing sometime. Anyway, we already have this reputation, people often stick to their first impression.

I already feel like some big crate were living in the past (I don't need to name them) and I hope there will not limit themselves just for few use cases. Of course, I don't encourage to always use the last release, but I think using 3-6 months release is perfectly acceptable, for me these are 3 versions of rustc, stable, beta (who use beta :p) and nightly. And we should have a nice error message that just ask to update rustc.

**Comment by kbknapp** [comment]
Here's a hacky PoC cargo-subcommand to generate a "quick" list of the MSRV by crate version for a particular crate. The idea being maintainers could list this data in their README and allow downstream consumers to make a more informed decision about upgrades and such.

https://github.com/kbknapp/cargo-msrv-table

**Comment by DianaNites** [comment]
Crates could specify it in Cargo.toml as package metadata? `cargo metadata` gives manifest paths to dependencies, so tool only needs to read manifests. Would Just Work among varying dependency versions, and where not specified the tool could just try compiling it based on your selected MSRV?

**Comment by est31** [comment]
@kbknapp nice tool! Some suggestions:

* Usage of bisection instead of linear search
* Maybe incorporating the versions found out for earlier crate releases? Usually it should be close, so maybe test the version itself as well as the version before and behind before starting bisection.
* Right now cargo doesn't respect MSRV during resolution, so your results will show too recent MSRVs. One can patch cargo's resolver to treat cargo releases with an MSRV prior to a certain version as yanked, as long as you have MSRV info for those crates.
* I recommend building a database instead of letting crate maintainers do it. This way, the new feature requires no buy in from the maintainers who are often even hostile to the MSRV concept (yet their code is still being relied upon). That database can be filled by automated tools running in the cloud or something.

**Comment by kbknapp** [comment]
@est31 great suggestions! The tool was meant to be more a proof of concept than a full blown project. If someone wants to foster that command into a real solution I'd be happy to help along where I can and also grant full access.

I agree some sort of database would be more usable, even if it was some sort of combination of local/"cooperative" crates instead of a full search. However, that also adds a lot of complexity, and potentially cost to such project.
